### PR TITLE
Allow DAPR apps without a http endpoint to function correctly

### LINKED ISSUE #3687: How can I add project without endpoints with Dapr sidecar?
Hello!

This used to work in preview 4 but stopped in preview 5.

I have a project which does not have any endpoints. 
It uses SDK to monitor storage queue (no Dapr bindings), it doesn't use any Dapr components, it just uses Dapr ActorProxy to invoke methods on actor instances.
Actors are hosted by a different project.

I can run this project with Dapr sidecar from command line from within project's directory like this (no app port):
`dapr run --app-id cloud-runs-controller -- dotnet run`

In preview 5, when I add the project to Aspire host, its Dapr sidecar fails to start with error

```
Failed to create resource package-event-processor-dapr-cli
System.InvalidOperationException: The endpoint `http` is not allocated for the resource `package-event-processor`.
   at Aspire.Hosting.ApplicationModel.EndpointReference.get_AllocatedEndpoint() in /_/src/Aspire.Hosting/ApplicationModel/EndpointReference.cs:line 96
   at Aspire.Hosting.ApplicationModel.EndpointReference.get_Host() in /_/src/Aspire.Hosting/ApplicationModel/EndpointReference.cs:line 78
   at Aspire.Hosting.Dapr.DaprDistributedApplicationLifecycleHook.<>c__DisplayClass6_1.<BeforeStartAsync>b__9(IList`1 updatedArgs) in /_/src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:line 194
   at Aspire.Hosting.ApplicationModel.CommandLineArgsCallbackAnnotation.<>c__DisplayClass1_0.<.ctor>b__0(CommandLineArgsCallbackContext c) in /_/src/Aspire.Hosting/ApplicationModel/CommandLineArgsCallbackAnnotation.cs:line 32
   at Aspire.Hosting.Dcp.ApplicationExecutor.CreateExecutableAsync(AppResource er, ILogger resourceLogger, CancellationToken cancellationToken) in /_/src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:line 1194
   at Aspire.Hosting.Dcp.ApplicationExecutor.<CreateExecutablesAsync>g__CreateExecutableAsyncCore|53_0(AppResource cr, CancellationToken cancellationToken) in /_/src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:line 1134
```

Code that adds the project to Apire host:

``` CSharp
builder.AddProject<Projects.PackageEventProcessor>("package-event-processor")
    .WithDaprSidecar(new DaprSidecarOptions
    {
        AppId = "cloud-runs-controller",
        ResourcesPaths = ImmutableHashSet.Create(@"..\resources"), // technically not needed

        // following is ok for local development with minimal logging
        EnableApiLogging = false,
        EnableAppHealthCheck = false,
        EnableProfiling = false,
        LogLevel = "warn",
        AppHealthProbeInterval = 60,
    });
```

### PR DESCRIPTION
**Pull request by oising**
## Description

The DAPR hosting `WithDaprSideCar()` expects the application to always expose a http endpoint. In the DAPR world, this is not the expectation. Only apps that expect to work with input bindings or subscribe to a pubsub topic (i.e. receive messages or events) are expected to have an http endpoint.   

It seems a regression was introduced where an endpoint validation was previously a null check, and now it requires an `IsAllocated` boolean test. I also simplified the tests into patterns and unified the three checks across all three `if` statements checking appPort, appChannelAddress and appProtocol respectively. 

I added a third "service C" to the playground app to validate that a service without an endpoint is successfully launched.

Unit tests (to come?) -- but they look pretty arcane. Might need some pointers.

Fixes #3687

## Checklist

- Is this feature complete?
  - [ ] Yes. Ready to ship.
  - [x] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [x] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

### CODE REVIEW

**Review by philliphoff — APPROVED** [review]
Looks ok to me.

File: playground/dapr/ServiceC/Program.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            await dapr.WaitForSidecarAsync(stoppingToken);
+
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                var state = await dapr.GetStateAsync<string?>(
+                    "statestore", "key", cancellationToken: stoppingToken);
+
+                Console.WriteLine($"State: {state ?? "<null>"}");
```
Comment (davidfowl) [inline playground/dapr/ServiceC/Program.cs:35]: ILogger?

**Comment by oising** [comment]
@davidfowl -- will this make next release? It's a blocker for us. Thx!

**Comment by davidfowl** [comment]
It could, there's still time, my big concern is that we have no automated test coverage for dapr at the moment.

**Comment by davidfowl** [comment]
Spoke to the team and since we have little dapr test coverage at the moment I will merge this into main so that we can do more testing. Worst case is that we can back port to 9.0.1 if it doesn't make this release.

**Comment by oising** [comment]
> Spoke to the team and since we have little dapr test coverage at the moment I will merge this into main so that we can do more testing. Worst case is that we can back port to 9.0.1 if it doesn't make this release.

I did add extend the playground app for it to verify, but yeah, I wish I could help more.

**Comment by davidfowl** [comment]
Can you help verify 9.1.0 builds?

**Comment by oising** [comment]
> Can you help verify 9.1.0 builds?

Yes -- what do you need me to do?

**Comment by davidfowl** [comment]
Verify that there are no regressions in using it with dapr.

**Pull request merged**